haystack/management/commands/update_index.py

```diff
@@ -57,64 +57,47 @@ def update_worker(args):
     backend = haystack_connections[using].get_backend()
 
     qs = index.build_queryset(start_date=start_date, end_date=end_date)
-    do_update(backend, index, qs, start, end, total, verbosity, commit, max_retries)
+    do_update(backend, index, qs, start, end, verbosity, commit, max_retries)
     return args
 
 
-def do_update(backend, index, qs, start, end, total, verbosity=1, commit=True,
-              max_retries=DEFAULT_MAX_RETRIES, last_max_pk=None):
-
-    # Get a clone of the QuerySet so that the cache doesn't bloat up
-    # in memory. Useful when reindexing large amounts of data.
-    small_cache_qs = qs.all()
-
-    # If we got the max seen PK from last batch, use it to restrict the qs
-    # to values above; this optimises the query for Postgres as not to
-    # devolve into multi-second run time at large offsets.
-    if last_max_pk is not None:
-        current_qs = small_cache_qs.filter(pk__gt=last_max_pk)[:end - start]
-    else:
-        current_qs = small_cache_qs[start:end]
-
-    # Remember maximum PK seen so far
-    max_pk = None
-    current_qs = list(current_qs)
-    if current_qs:
-        max_pk = current_qs[-1].pk
+def do_update(backend, index, qs, start_pk, end_pk, verbosity=1, commit=True,
+              max_retries=DEFAULT_MAX_RETRIES):
+    current_qs = qs.filter(pk__range=(start_pk, end_pk)).order_by()
 
     is_parent_process = hasattr(os, 'getppid') and os.getpid() == os.getppid()
 
     if verbosity >= 2:
         if is_parent_process:
-            print("  indexed %s - %d of %d." % (start + 1, end, total))
+            print(u"  indexing %s - %s." % (start_pk, end_pk))
```

Comment:
Strictly speaking the `u` prefix is unnecessary since we have `from __future__ import unicode_literals` in this file

Reply:
While we're in here, I'm also wondering whether we should just switch to logging for anyone who is calling this command from a context other than a normal shell. Maybe that's a ticket for followup work.